### PR TITLE
Check the path for spaces before sending. Avoids filenames to be inte…

### LINKED ISSUE #56109: Run active file [with spaces in name] in active terminal
Issue Type: <b>Bug</b>

I checked the issue reproes in the latest code version.

1 - Open a file that contains whitespaces in name. (In this example "AA Generator.py")
2 - Execute "Run Active File In Active Terminal"

Expected result: The file is executed, either quoted :
- PS C:\Users\luiszun\git\vscode> "c:\Users\luiszun\Desktop\AA Generator.py"

Actual result: The incomplete-filename is executed with the rest interpreted as a parameter.

PS C:\Users\luiszun\git\vscode> c:\Users\luiszun\Desktop\AA Generator.py
c:\Users\luiszun\Desktop\AA : The term 'c:\Users\luiszun\Desktop\AA' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included,
verify that the path is correct and try again.
At line:1 char:1

VS Code version: code-oss-dev 1.27.0 (Commit unknown, Date unknown)
OS version: Windows_NT x64 10.0.17134

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz (8 x 2712)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Memory (System)|15.79GB (3.75GB free)|
|Process Argv|C:\Users\luiszun\git\vscode\.build\electron\Code - OSS.exe . --remote-debugging-port=9222 --no-first-run --no-default-browser-check --inspect=5875|
|Screen Reader|no|
|VM|0%|

</details>Extensions: none

### PR DESCRIPTION
**Pull request by luiszun**
Check the path for spaces before sending. Avoids filenames to be interpreted as file and arguments

Closes #56109

### CODE REVIEW

**Review by ctrlaltvikas — APPROVED** [review]

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts [hunk 1]
```diff
@@ -638,7 +638,11 @@ export class RunActiveFileInTerminalAction extends Action {
 			this.notificationService.warn(nls.localize('workbench.action.terminal.runActiveFile.noFile', 'Only files on disk can be run in the terminal'));
 			return TPromise.as(void 0);
 		}
-		instance.sendText(uri.fsPath, true);
+		let uriPath: string = uri.fsPath;
+		if (uriPath.indexOf(' ') !== -1) {
```
[the remark this answers is not in the document]
Comment (luiszun) [reply src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts]: The " character is not allowed in Windows.
I thought it was the same in Unixes, but you're right! Some shells will not allow it, but at least the major filesystems allow it (ext3, ext4). I'll get educated on Mac, and I'll change the PR.

Thanks!

[the remark this answers is not in the document]
Comment (Tyriar) [reply src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts]: Mac allows it, just testing Linux and treating mac the same should be enough 👍

[the remark this answers is not in the document]
Comment (alexr00) [reply src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts]: Looks like on Linux space can also be dealt with by surrounding with single quotes.

**Comment by alexr00** [comment]
I tested using WSL and this path: ~tes"t/test script.bash

**Comment by luiszun** [comment]
I thought of doing the single quote thing, but I found out that some filesystems (ext4 being a very widely used one) allow all characters except '\' and null. Though I doubt escaping non-printable characters is a practical solution. I think that for all practical purposes, the single quote suffices. Just wanted to drop this piece of data :)

File: src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts [hunk 2]
```diff
@@ -653,7 +654,15 @@ export class RunActiveFileInTerminalAction extends Action {
 			this.notificationService.warn(nls.localize('workbench.action.terminal.runActiveFile.noFile', 'Only files on disk can be run in the terminal'));
 			return TPromise.as(void 0);
 		}
-		instance.sendText(uri.fsPath, true);
+		let uriPath: string = uri.fsPath;
+		if ((uriPath.indexOf(' ') !== -1) || (uriPath.indexOf('"') !== -1)) {
+			if (isWindows) {
+				uriPath = '"' + uriPath + '"';
+			} else {
+				uriPath = '\'' + uriPath + '\'';
```
Comment (Tyriar) [inline src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts]: Escaping on macOS and Linux would probably be better done by escaping each space, this is more typical than quoting the whole path. This is what happens when you drag a file in that contains spaces: `/Users/daimms/Desktop/Screen\ Shot\ 2018-09-14\ at\ 12.39.14\ PM.png`

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts [hunk 3]
```diff
@@ -653,7 +654,19 @@ export class RunActiveFileInTerminalAction extends Action {
...
+		let uriPath: string = uri.fsPath;
+		const hasSpace = uriPath.indexOf(' ') !== -1;
+		if (hasSpace && isWindows) {
+			uriPath = '"' + uriPath + '"';
+		} else if (!isWindows) {
+			const hasDoubleQuote = uriPath.indexOf('"') !== -1;
+			if (!hasSpace && hasDoubleQuote) {
+				uriPath = '\'' + uriPath + '\'';
+			} else if (hasSpace) {
+				uriPath = uriPath.replace(' ', '\\ ');
```
Comment (Tyriar) [inline src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts]: `.replace` only replaces the first instance.

I think we also need to handle `\\` -> `\\\\`

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Comment by Tyriar** [comment]
@luiszun could you please sign the CLA mentioned in https://github.com/Microsoft/vscode/pull/56966#issuecomment-422596291 so we can get this merged?

**Review by Tyriar — APPROVED** [review]

**Comment by Tyriar** [comment]
Looks good to me, just waiting on the CLA

**Comment by luiszun** [comment]
Supposedly I didn't have to sign the CLA if I linked my Microsoft corp account to this profile.

Anyway, just signed it. Sorry for the delay

**Pull request merged**